numba_dpex/core/kernel_interface/arg_pack_unpacker.py

```diff
@@ -0,0 +1,291 @@
+            total_size: Total number of items in the array.
+            shape: Shape of the array.
+            ndim: Total number of dimensions.
+            itemsize: Size of each item.
+            strides: Stride of the array.
+            dtype: Dtype of the array.
+        """
+        try:
+            usm_mem = dpctl_mem.as_usm_memory(obj)
+        except Exception:
+            logging.exception(
+                "array-like object does not implement the SUAI protocol."
+            )
+            raise SUAIProtocolError(self._pyfunc_name, obj)
+
+        shape = obj.__sycl_usm_array_interface__["shape"]
+        total_size = np.prod(obj.__sycl_usm_array_interface__["shape"])
+        ndim = len(obj.__sycl_usm_array_interface__["shape"])
+        itemsize = np.dtype(
+            obj.__sycl_usm_array_interface__["typestr"]
+        ).itemsize
+        dtype = np.dtype(obj.__sycl_usm_array_interface__["typestr"])
+        strides = obj.__sycl_usm_array_interface__["strides"]
+
+        if strides is None:
+            strides = [1] * ndim
+            for i in reversed(range(1, ndim)):
+                strides[i - 1] = strides[i] * shape[i]
+            strides = tuple(strides)
+
+        return usm_mem, total_size, shape, ndim, itemsize, strides, dtype
+
+    def _unpack_array_helper(self, size, itemsize, buf, shape, strides, ndim):
+        """
+        Implements the unpacking logic for array arguments.
+
+        TODO: Add more detail
+
+        Args:
+            size: Total number of elements in the array.
+            itemsize: Size in bytes of each element in the array.
+            buf: The pointer to the memory.
+            shape: The shape of the array.
+            ndim: Number of dimension.
+
+        Returns:
+            A list a ctype value for each array attribute argument
+        """
+        unpacked_array_attrs = []
+
+        # meminfo (FIXME: should be removed and the USMArrayType modified once
+        # NumPy support is removed)
+        unpacked_array_attrs.append(ctypes.c_size_t(0))
+        # meminfo (FIXME: Evaluate if the attribute should be removed and the
+        # USMArrayType modified once NumPy support is removed)
+        unpacked_array_attrs.append(ctypes.c_size_t(0))
+        unpacked_array_attrs.append(ctypes.c_longlong(size))
+        unpacked_array_attrs.append(ctypes.c_longlong(itemsize))
+        unpacked_array_attrs.append(buf)
+        for ax in range(ndim):
+            unpacked_array_attrs.append(ctypes.c_longlong(shape[ax]))
+        for ax in range(ndim):
+            unpacked_array_attrs.append(ctypes.c_longlong(strides[ax]))
+
+        return unpacked_array_attrs
+
+    def _unpack_usm_array(self, val):
+        (
+            usm_mem,
+            total_size,
+            shape,
+            ndim,
+            itemsize,
+            strides,
+            dtype,
+        ) = self._get_info_from_suai(val)
+
+        return self._unpack_array_helper(
```

Comment:
Not clear why we need `_unpack_array_helper`. `unpack_usm_array` is basically doing almost everything what `_unpack_array_helper` is doing.

Reply:
I agree. It is one of the things that I will clean up with the removal of NumPy support from kernels.